src/libtls/tls_server.c

```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+		reader->destroy(reader);
+		if (!found)
+		{
+			ht->remove_at(ht, enumerator);
+			free(v);
+		}
+	}
+	enumerator->destroy(enumerator);
+	return ht;
+}
+
 /**
  * Find a cipher suite and a server key
  */
 static bool select_suite_and_key(private_tls_server_t *this,
 								 tls_cipher_suite_t *suites, int count)
 {
+	hashtable_t *key_types;
+	enumerator_t *enumerator;
+	tls_version_t version_min, version_max;
 	private_key_t *key;
-	key_type_t type;
+	key_type_t type = 0, *v;
+	int *k;
 
-	key = lib->credmgr->get_private(lib->credmgr, KEY_ANY, this->server,
-									this->server_auth);
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	key_types = create_common_key_types(this->hashsig, version_min, version_max);
+	if (!key_types)
+	{
+		DBG1(DBG_TLS, "no common signature algorithm found");
+		return FALSE;
+	}
+	enumerator = key_types->create_enumerator(key_types);
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		key = lib->credmgr->get_private(lib->credmgr, *v, this->server,
+										this->server_auth);
+		if (key)
+		{
+			type = *v;
+			break;
+		}
+		key_types->remove_at(key_types, enumerator);
+		free(v);
+	}
+	enumerator->destroy(enumerator);
 	if (!key)
 	{
 		DBG1(DBG_TLS, "no usable TLS server certificate found for '%Y'",
 			 this->server);
+		key_types->destroy(key_types);
 		return FALSE;
 	}
 
-	if (this->tls->get_version_max(this->tls) >= TLS_1_3)
+	if (version_max >= TLS_1_3)
 	{
-		/* currently no support to derive key based on client supported
-		 * signature schemes */
 		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 												  		count, KEY_ANY);
-		if (!this->suite)
-		{
-			DBG1(DBG_TLS, "received cipher suites unacceptable");
-			return FALSE;
-		}
 	}
 	else
 	{
-		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
-												  		count,
-												  		key->get_type(key));
-		if (!this->suite)
-		{	/* no match for this key, try to find another type */
-			if (key->get_type(key) == KEY_ECDSA)
-			{
-				type = KEY_RSA;
-			}
-			else
-			{
-				type = KEY_ECDSA;
-			}
-			key->destroy(key);
-
+		if (type)
+		{
 			this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 															count, type);
-			if (!this->suite)
-			{
-				DBG1(DBG_TLS, "received cipher suites unacceptable");
-				return FALSE;
-			}
-			this->server_auth->destroy(this->server_auth);
-			this->server_auth = auth_cfg_create();
-			key = lib->credmgr->get_private(lib->credmgr, type, this->server,
-											this->server_auth);
-			if (!key)
-			{
-				DBG1(DBG_TLS, "received cipher suites unacceptable");
-				return FALSE;
+		}
+		if (!this->suite)
+		{
+			enumerator = key_types->create_enumerator(key_types);
+			while (enumerator->enumerate(enumerator, &k, &v))
+			{	/* find a cipher suite that supports the key */
+				type = *v;
+				key_types->remove_at(key_types, enumerator);
+				free(v);
+
+				if (!key)
```

Comment:
This should always be the case at this point (there is a `!key` check above that returns from the function). To keep the order the same, just free the existing key (and I guess auth_cfg), do the lookup with the new type (continue if none is found) and then look for a suite. Then go to the next type and free the key again etc. If you use an array you could even get rid of the `!this->suite` check and just do `while (!this->suite && array_remove...)` to check another key type.